File: ibm-ems-springboot/src/main/java/com/ibm/ems/service/ProjectService.java
package com.ibm.ems.service;

import java.util.List;

import com.ibm.ems.dto.employee.EmployeeResponse;
import com.ibm.ems.dto.project.AssignEmployeeRequest;
import com.ibm.ems.dto.project.ProjectRequest;
import com.ibm.ems.dto.project.ProjectResponse;
import com.ibm.ems.model.ProjectStatus;

public interface ProjectService {
	List<ProjectResponse> getAllProjects(ProjectStatus status);

	ProjectResponse getProjectById(String id);

	ProjectResponse createProject(ProjectRequest request);

	ProjectResponse updateProject(String id, ProjectRequest request);

	void deleteProject(String id);

	void assignEmployee(String projectId, AssignEmployeeRequest req);

	void removeEmployee(String projectId, String employeeId);

	List<EmployeeResponse> getProjectEmployees(String projectId);

	List<ProjectResponse> getEmployeeProjects(String employeeId);
}

File: src/app/service/project/project-service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment } from '../../../environments/environment';
import { get } from 'http';
import { EmployeeResponse } from '../employee/employee-service';
export interface ProjectRequest {
  name: string;
  description: string;

  startDate: string; // yyyy-MM-dd
  endDate: string;   // yyyy-MM-dd

  status: 'active'|'inactive';
}
export interface ProjectResponse extends ProjectRequest{
  id: string;
}

export interface AssignEmployeeRequest {
  employeeId: string;
  projectRole: string;
  assignedDate?: string; // yyyy-MM-dd
}

const BASE_URL=`${environment.baseUrl}/api/v1/projects`
@Injectable({
  providedIn: 'root',
})
export class ProjectService {
  constructor(
    private readonly client:HttpClient
  ){}

  getAllProjects=(status?:'active'|'inactive')=>{
    const path=`${BASE_URL}`
    let param=new HttpParams()
    if(status){
      param=param.append("status",status)
    }
    return this.client.get<ProjectResponse[]>(path,{
      params:param
    });
  }
  getProjectByID=(id:string)=>{
    const path=`${BASE_URL}/${id}`
    return this.client.get<ProjectResponse>(path)
  }
  addProject=(reqBody:ProjectRequest)=>{
    const path=`${BASE_URL}`
    return this.client.post<ProjectResponse>(path,reqBody)
  }
  updateProject=(id:string,reqBody:ProjectRequest)=>{
    const path=`${BASE_URL}/${id}`
    return this.client.put<ProjectResponse>(path,reqBody)
  }
  deleteProject=(id:string)=>{
    const path=`${BASE_URL}/${id}`
    return this.client.delete<void>(path)
  }
  assignEmployee=(id:string,reqBody:AssignEmployeeRequest)=>{
    const path=`${BASE_URL}/${id}/employees`
    return this.client.post<void>(path,reqBody)
  }
  removeEmployeeFromProject=(project_id:string,employee_id:string)=>{
    const path=`${BASE_URL}/${project_id}/employees/${employee_id}`
    return this.client.delete<void>(path)
  }
  getProjectEmployees=(id:string)=>{
    const path=`${BASE_URL}/${id}/employees`
    return this.client.get<EmployeeResponse[]>(path)
  }
  getEmployeeAssignedProjects=(id:string)=>{
    const path=`${BASE_URL}/employee/${id}`
    return this.client.get<ProjectResponse[]>(path)
  }
}
